mgr/dashboard: Improvements in cluster > OSDs

aaSharma14 opened the pull request:
Made osd-id the first column of the table instead of Host.

Fixes:https://tracker.ceph.com/issues/47478
Signed-off-by:Aashish Sharma <aasharma@redhat.com>

<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:
```diff
@@ -378,11 +384,19 @@ export class OsdListComponent extends ListWithDetails implements OnInit {
     });
   }
 
+  get_osd_device_count(id: number) {
```
Can you add a unit test for this new behavior?

votdev replied:
I think this is a better solution, also because you are calling an observable within this function.
```suggestion
  get_osd_device_count(osd: object) {
    this.osdService.getDevices(osd.id).subscribe((data: any[]) => {
      osd.device_count = data.length;
    });
  }
```

alfonsomthd commented:
jenkins test dashboard

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:
```diff
@@ -67,6 +67,7 @@ export class OsdListComponent extends ListWithDetails implements OnInit {
 
   selection = new CdTableSelection();
   osds: any[] = [];
+  device_count: any;
```
Not needed when using the suggested `get_osd_device_count` function. Remove it please.
```suggestion
```

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-list/osd-list.component.ts:
```diff
@@ -378,11 +384,19 @@ export class OsdListComponent extends ListWithDetails implements OnInit {
         osd.stats.usage = osd.stats.stat_bytes_used / osd.stats.stat_bytes;
         osd.cdIsBinary = true;
+        this.get_osd_device_count(osd.id);
+        osd.device_count = this.device_count;
```
Not needed. Objects can be modified within JS functions when they are passed as function parameters.
```suggestion
```

nizamial09 commented:
@aaSharma14 Could you please add the screenshot of the change?

aaSharma14 commented:
> @aaSharma14 Could you please add the screenshot of the change?

Added the screenshot.Thanks :)

aaSharma14 commented:
jenkins test make check

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -69,6 +70,17 @@ def list(self):
+            device_paths = metadata.find('device_paths')
+            devices = metadata[device_ids:device_paths-3]
+            devices = devices[14:len(devices)-1]
+            if devices == "":
```
Following this: https://docs.python.org/3/library/stdtypes.html#truth-value-testing
You can simplify it:
```
osd['device_count'] = 0
if devices:
    device_list = devices.split(',')
    osd['device_count'] = len(device_list)
```

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -69,6 +70,16 @@ def list(self):
             osd['stats'] = {}
             osd['stats_history'] = {}
             osd_spec = str(osd_id)
+            metadata = str(mgr._ceph_get_metadata('osd', osd_spec))  # pylint: disable = W0212
```
```suggestion
            metadata = str(mgr.get_metadata('osd', osd_spec))
```

aaSharma14 replied:
Both methods have the same output, however using mgr.get_metadata was giving an unusual 'Key Error' in the test_cephfs.py, so I went on using the _ceph_get_metadata method.

alfonsomthd replied:
Then we should investigate the cause of the failure and adapt the test accordingly.
We should call the public counterpart of the protected method.

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -44,7 +44,8 @@ def osd_task(name, metadata, wait_for=2.0):
 @ControllerDoc('OSD management API', 'OSD')
 class Osd(RESTController):
-    def list(self):
+    @handle_send_command_error('osd')
```
Is this decorator still needed?

aaSharma14 replied:
It was as per the previous implementation, deleted it now, Thanks :)

aaSharma14 commented:
jenkins test make check

alfonsomthd commented:
@aaSharma14 I've tested it and does not work as expected: OSD device count = 0 when it should be 2:

![Screenshot from 2020-10-15 18-20-14](https://user-images.githubusercontent.com/6780162/96158896-4ec18500-0f14-11eb-9778-93add4a3b029.png)

Here you can see that the osd has 2 devices:
![Screenshot from 2020-10-15 18-20-38](https://user-images.githubusercontent.com/6780162/96158963-6567dc00-0f14-11eb-818e-6c0596059603.png)

A fix suggestion for your consideration:
```
            metadata = mgr.get_metadata('osd', osd_spec)
            osd['device_count'] = 0
            if metadata and 'devices' in metadata:
                osd['device_count'] = len(metadata['devices'].split(','))
```

aaSharma14 commented:
jenkins test make check

aaSharma14 commented:
> @aaSharma14 I've tested it and does not work as expected: OSD device count = 0 when it should be 2:
> 
> ![Screenshot from 2020-10-15 18-20-14](https://user-images.githubusercontent.com/6780162/96158896-4ec18500-0f14-11eb-9778-93add4a3b029.png)
> 
> Here you can see that the osd has 2 devices:
> ![Screenshot from 2020-10-15 18-20-38](https://user-images.githubusercontent.com/6780162/96158963-6567dc00-0f14-11eb-818e-6c0596059603.png)
> 
> A fix suggestion for your consideration:
> 
> ```
>             metadata = mgr.get_metadata('osd', osd_spec)
>             osd['device_count'] = 0
>             if metadata and 'devices' in metadata:
>                 osd['device_count'] = len(metadata['devices'].split(','))
> ```
> 
> Also, IMHO "Device Count" column needs a little bit more of width (or maybe the text can be shortened).

Thanks @alfonsomthd , I have made some modifications and updated the PR, Can you please do a retest and confirm if its working correctly now.!

alfonsomthd reviewed (approved):
Tested locally: LGTM

epuertat commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -68,6 +68,12 @@ def list(self):
             osd['stats'] = {}
             osd['stats_history'] = {}
             osd_spec = str(osd_id)
+            metadata = mgr.get_metadata('osd', osd_spec)
```
Could you please measure [how long this `list()` method takes](https://stackoverflow.com/a/27737385/13173750) with: 1, 2, 4, 8 and 16 OSDs. You can easily do that in vstart (`OSD=N` env var setting). It's also be interesting to see how the CPU reacts during this request (you'll probably better launch this from REST via curl CLI).

epuertat replied:
With the data that @aaSharma14 obtained from running this with OSD count {1,2,4,6,8,16}, as expected, it grows linearly (not many samples but enough to confirm it). Despite not resulting in huge numbers when extrapolated to OSD=100 or even 1000 (35 ms or 350 ms), the number of API calls (O(N)) might cause issues with the processing queues at ceph-mgr.

![image](https://user-images.githubusercontent.com/37327689/97331598-dba40100-1879-11eb-966b-fd3c0d07bb6f.png)

Additionally, this `mgr.get_metadata()` call is only used for retrieved the number of devices of each OSD, which is a rarely changing data.

Solutions:
* To find a single ceph-mgr API call returning this info aggregated for all OSDs.
* Or to introduce pagination, or at least only retrieving separately the currently displayed OSD information, and the device count (only for the OSDs displayed). As a side effect of this, the Device Count column couldn't be sortable/searchable anymore (we don't have all OSD's device count to sort that or search).

epuertat replied:
This PR meant to fix 2 issues reported in the linked tracker (@pcuzner please check this):

> 1. The table provides a list of OSDs and yet the first column isn’t the osd id it’s the host
>   1.1. Put the osd id first
> 2. Each osd can span multiple devices (and does for spinners), but to understand how many devices support an OSD you have to go to the properties of each OSD..
>   2.1. Provide the count of devices an OSD uses on each row to provide a quick way for the configuration to be checked.

The 1st one is tackled, while the second 2 is "feasible" but would impact scalability: OSD endpoint is now O(1) (as long as `mgr.get('osd_map')` is O(1) which is probably an 'over-assumption' given CPython interfacing overhead) and for retrieving this information, it's necessary to poll `mgr.metadata()` for every OSD, which is O(N) for a periodically polled endpoint...

My suggestion is to keep the current behaviour: leave the device count in the OSD Details sub-table, and only poll for that information when a user wants to expand details for that OSD. Additionally, the device count is 'low-entropy' data (will never change once an OSD is created, and most, if no all, OSDs will have the same device count).

aaSharma14 replied:
Thanks @epuertat . I have removed the device count feature as discussed and this PR is now only intended for :
The table provides a list of OSDs and yet the first column isn’t the osd id it’s the host
1.1. Put the osd id first

nizamial09 replied:
@aaSharma14 Could you update the description with a screenshot?

epuertat reviewed (dismissed):
I have a big concern with the newly added `get_metadata` method. This will probably happen to other `list()` methods too, but we should be extremely careful before adding more bottlenecks to the dashboard code.
